tests/tests.rs

```diff
@@ -0,0 +1,200 @@
+use std::rc::Rc;
+
+use candid::{CandidType, Decode, Encode, Nat};
+use evm_rpc::*;
+use ic_base_types::{CanisterId, PrincipalId};
+use ic_ic00_types::BoundedVec;
+use ic_state_machine_tests::{CanisterSettingsArgs, StateMachine, StateMachineBuilder, WasmResult};
+use ic_test_utilities_load_wasm::load_wasm;
+use serde::de::DeserializeOwned;
+
+const DEFAULT_CALLER_TEST_ID: u64 = 10352385;
+const DEFAULT_CONTROLLER_TEST_ID: u64 = 10352386;
+
+fn evm_rpc_wasm() -> Vec<u8> {
+    load_wasm(std::env::var("CARGO_MANIFEST_DIR").unwrap(), "evm_rpc", &[])
+}
+
+fn assert_reply(result: WasmResult) -> Vec<u8> {
+    match result {
+        WasmResult::Reply(bytes) => bytes,
+        result => {
+            panic!("Expected a successful reply, got {}", result)
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct EvmRpcSetup {
+    pub env: Rc<StateMachine>,
+    pub caller: PrincipalId,
+    pub controller: PrincipalId,
+    pub evm_rpc_id: CanisterId,
+}
+
+impl Default for EvmRpcSetup {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl EvmRpcSetup {
+    pub fn new() -> Self {
+        let env = Rc::new(
+            StateMachineBuilder::new()
+                .with_default_canister_range()
+                .build(),
+        );
+
+        let controller = PrincipalId::new_user_test_id(DEFAULT_CONTROLLER_TEST_ID);
+        let evm_rpc_id = env.create_canister(Some({
+            let mut args: CanisterSettingsArgs = Default::default();
+            args.controllers = Some(BoundedVec::new(vec![controller]));
+            args
+        }));
+        env.install_existing_canister(evm_rpc_id, evm_rpc_wasm(), Encode!(&()).unwrap())
+            .unwrap();
+
+        let caller = PrincipalId::new_user_test_id(DEFAULT_CALLER_TEST_ID);
+
+        Self {
+            env,
+            caller,
+            controller,
+            evm_rpc_id,
+        }
+    }
+
+    pub fn as_controller(&self) -> Self {
```

Comment:
Since the function is only used in testing, wouldn't it be easier to simply use a mutable reference where you can set the controller (and caller...)?

Reply:
It's also possible to change the caller/controller via mutable reference. This is essentially a shorthand for cloning and changing the caller, since this will be used in most of the state machine tests.